### PR TITLE
Fixed assignment of u_old_old_old in ExplicitDynamicsContactConstraint

### LINKED ISSUE #27883: Issue with explicit contact dynamics
### Bug Description

When running a simulation using explicit contact dynamics with a central difference time integrator, the simulation will fail due to an inverted Jacobian error once contact occurs.

The issue is only caught when the impact velocity is higher than that applied in the current regression tests. 

### Steps to Reproduce

Here I modified the `settlement.i` input file found in `modules/contact/test/tests/explicit_dynamics/settlement.i`. 
This test drops a block onto a rigid plate. 
This is modified to drop the block from a 0.5m height, as opposed to the 0.0001 height from the original test. 

```
# One element test to test the central difference time integrator in 3D.
[GlobalParams]
    displacements = 'disp_x disp_y disp_z'
    volumetric_locking_correction = true
[]

[Mesh]
    [block_one]
        type = GeneratedMeshGenerator
        dim = 3
        nx = 3
        ny = 3
        nz = 3
        xmin = 4.5
        xmax = 5.5
        ymin = 4.5
        ymax = 5.5
        zmin = 0.5
        zmax = 1.5
        boundary_name_prefix = 'ball'
    []
    [block_two]
        type = GeneratedMeshGenerator
        dim = 3
        nx = 2
        ny = 2
        nz = 2
        xmin = 0.0
        xmax = 10
        ymin = 0.0
        ymax = 10
        zmin = -2
        zmax = 0
        boundary_name_prefix = 'base'
        boundary_id_offset = 10
    []
    [block_one_id]
        type = SubdomainIDGenerator
        input = block_one
        subdomain_id = 1
    []
    [block_two_id]
        type = SubdomainIDGenerator
        input = block_two
        subdomain_id = 2
    []
    [combine]
        type = MeshCollectionGenerator
        inputs = ' block_one_id block_two_id'
    []
[]

[AuxVariables]
    [penetration]
    []
[]

[AuxKernels]
    [penetration]
        type = PenetrationAux
        variable = penetration
        boundary = ball_back
        paired_boundary = base_front
        quantity = distance
    []
[]

[Variables]
    [disp_x]
    []
    [disp_y]
    []
    [disp_z]
    []
[]

[AuxVariables]
    [gap_rate]
    []
    [vel_x]
    []
    [accel_x]
    []
    [vel_y]
    []
    [accel_y]
    []
    [vel_z]
    []
    [accel_z]
    []
    [stress_zz]
        family = MONOMIAL
        order = CONSTANT
    []
    [strain_zz]
        family = MONOMIAL
        order = CONSTANT
    []
    [kinetic_energy_one]
        order = CONSTANT
        family = MONOMIAL
    []
    [elastic_energy_one]
        order = CONSTANT
        family = MONOMIAL
    []
    [kinetic_energy_two]
        order = CONSTANT
        family = MONOMIAL
    []
    [elastic_energy_two]
        order = CONSTANT
        family = MONOMIAL
    []
[]

[AuxKernels]
    [stress_zz]
        type = RankTwoAux
        rank_two_tensor = stress
        index_i = 2
        index_j = 2
        variable = stress_zz
        execute_on = 'TIMESTEP_END'
    []
    [strain_zz]
        type = RankTwoAux
        rank_two_tensor = mechanical_strain
        index_i = 2
        index_j = 2
        variable = strain_zz
    []
    [accel_x]
        type = TestNewmarkTI
        variable = accel_x
        displacement = disp_x
        first = false
        execute_on = 'LINEAR TIMESTEP_BEGIN TIMESTEP_END'
    []
    [vel_x]
        type = TestNewmarkTI
        variable = vel_x
        displacement = disp_x
        execute_on = 'LINEAR TIMESTEP_BEGIN TIMESTEP_END'
    []
    [accel_y]
        type = TestNewmarkTI
        variable = accel_y
        displacement = disp_y
        first = false
        execute_on = 'LINEAR TIMESTEP_BEGIN TIMESTEP_END'
    []
    [vel_y]
        type = TestNewmarkTI
        variable = vel_y
        displacement = disp_x
        execute_on = 'LINEAR TIMESTEP_BEGIN TIMESTEP_END'
    []
    [accel_z]
        type = TestNewmarkTI
        variable = accel_z
        displacement = disp_z
        first = false
        execute_on = 'LINEAR TIMESTEP_BEGIN TIMESTEP_END'
    []
    [vel_z]
        type = TestNewmarkTI
        variable = vel_z
        displacement = disp_z
        execute_on = 'LINEAR TIMESTEP_BEGIN TIMESTEP_END'
    []
    [kinetic_energy_one]
        type = KineticEnergyAux
        block = '1'
        variable = kinetic_energy_one
        newmark_velocity_x = vel_x
        newmark_velocity_y = vel_y
        newmark_velocity_z = vel_z
        density = density
    []
    [elastic_energy_one]
        type = ElasticEnergyAux
        variable = elastic_energy_one
        block = '1'
    []
    [kinetic_energy_two]
        type = KineticEnergyAux
        block = '2'
        variable = kinetic_energy_two
        newmark_velocity_x = vel_x
        newmark_velocity_y = vel_y
        newmark_velocity_z = vel_z
        density = density
    []
    [elastic_energy_two]
        type = ElasticEnergyAux
        variable = elastic_energy_two
        block = '2'
    []
[]

[Kernels]
    [DynamicTensorMechanics]
        displacements = 'disp_x disp_y disp_z'
        stiffness_damping_coefficient = 1.0e-3
        generate_output = 'stress_zz strain_zz'
    []
    [inertia_x]
        type = InertialForce
        variable = disp_x
    []
    [inertia_y]
        type = InertialForce
        variable = disp_y
    []
    [inertia_z]
        type = InertialForce
        variable = disp_z
    []
    [gravity]
        type = Gravity
        variable = disp_z
        value = -98.10
    []
[]

[BCs]
    [x_front]
        type = DirichletBC
        variable = disp_x
        boundary = 'ball_front'
        preset = false
        value = 0.0
    []
    [y_front]
        type = DirichletBC
        variable = disp_y
        boundary = 'ball_front'
        preset = false
        value = 0.0
    []
    [x_fixed]
        type = DirichletBC
        variable = disp_x
        boundary = 'base_back'
        preset = false
        value = 0.0
    []
    [y_fixed]
        type = DirichletBC
        variable = disp_y
        boundary = 'base_back'
        preset = false
        value = 0.0
    []
    [z_fixed]
        type = DirichletBC
        variable = disp_z
        boundary = 'base_back'
        preset = false
        value = 0.0
    []
    [z_fixed_front]
        type = DirichletBC
        variable = disp_z
        boundary = 'base_front'
        preset = false
        value = 0.0
    []
[]

[ExplicitDynamicsContact]
    [my_contact]
        model = frictionless_balance
        primary = base_front
        secondary = ball_back
        vel_x = 'vel_x'
        vel_y = 'vel_y'
        vel_z = 'vel_z'
    []
[]

[Materials]
    [elasticity_tensor_block_one]
        type = ComputeIsotropicElasticityTensor
        youngs_modulus = 1e6
        poissons_ratio = 0.0
        block = 1
        outputs = 'exodus'
        output_properties = __all__
    []
    [elasticity_tensor_block_two]
        type = ComputeIsotropicElasticityTensor
        youngs_modulus = 1e10
        poissons_ratio = 0.0
        block = 2
        outputs = 'exodus'
        output_properties = __all__
    []
    [strain_block]
        type = ComputeFiniteStrain # ComputeIncrementalSmallStrain
        displacements = 'disp_x disp_y disp_z'
        implicit = false
    []
    [stress_block]
        type = ComputeFiniteStrainElasticStress
    []
    [density_one]
        type = GenericConstantMaterial
        prop_names = density
        prop_values = 1e1
        outputs = 'exodus'
        output_properties = 'density'
        block = '1'
    []
    [density_two]
        type = GenericConstantMaterial
        prop_names = density
        prop_values = 1e6
        outputs = 'exodus'
        output_properties = 'density'
        block = '2'
    []
    [wave_speed]
        type = WaveSpeed
        outputs = 'exodus'
        output_properties = 'wave_speed'
    []
[]

[Executioner]
    type = Transient
    start_time = -0.01
    end_time = 0.2
    dt = 1.0e-4
    timestep_tolerance = 1e-6

    [TimeIntegrator]
        type = CentralDifference
        solve_type = lumped
    []
[]

[Outputs]
    interval = 1
    exodus = true
    csv = true
    file_base = 'settlement_test'
[]

[Postprocessors]
    [accel_58z]
        type = NodalVariableValue
        nodeid = 1
        variable = accel_z
    []
    [vel_58z]
        type = NodalVariableValue
        nodeid = 1
        variable = vel_z
    []
    [critical_time_step]
        type = CriticalTimeStep
    []
    [contact_pressure_max]
        type = NodalExtremeValue
        variable = contact_pressure
        block = '1 2'
        value_type = max
    []
    [penetration_max]
        type = NodalExtremeValue
        variable = penetration
        block = '1 2'
        value_type = max
    []
    [total_kinetic_energy_one]
        type = ElementIntegralVariablePostprocessor
        variable = kinetic_energy_one
        block = '1'
    []
    [total_elastic_energy_one]
        type = ElementIntegralVariablePostprocessor
        variable = elastic_energy_one
        block = '1'
    []
    [total_kinetic_energy_two]
        type = ElementIntegralVariablePostprocessor
        variable = kinetic_energy_two
        block = '2'
    []
    [total_elastic_energy_two]
        type = ElementIntegralVariablePostprocessor
        variable = elastic_energy_two
        block = '2'
    []
[]
```

Running this file on the most recent `devel` commit will result in an inverted Jacobian error on contact. 

However, this error does not occur when reverting commit 149e6c3. After reversion, contact occurs without issue. This commit represents a change to the central difference formulation. It is unclear why these changes have a drastic effect on explicit contact. 

### Impact

I am unable to complete explicit contact simulations that have an appreciable impact velocity. 

### [Optional] Diagnostics

```
##################################################################################################
Influential Environment Variables

CC=mpicc
CC_FOR_BUILD=/Users/torrde/miniforge/envs/moose/bin/x86_64-apple-darwin13.4.0-clang
CFLAGS=-march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -isystem /Users/torrde/miniforge/envs/moose/include
CONDA_CHANNEL=https://conda.software.inl.gov/public
CONDA_DEFAULT_ENV=moose
CONDA_EXE=/Users/torrde/miniforge/bin/conda
CONDA_PREFIX=/Users/torrde/miniforge/envs/moose
CONDA_PREFIX_1=/Users/torrde/miniforge
CONDA_PROMPT_MODIFIER=(moose) 
CONDA_PYTHON_EXE=/Users/torrde/miniforge/bin/python
CONDA_SHLVL=2
CONDA_TOOLCHAIN_BUILD=x86_64-apple-darwin13.4.0
CONDA_TOOLCHAIN_HOST=x86_64-apple-darwin13.4.0
CPPFLAGS=-D_FORTIFY_SOURCE=2 -isystem /Users/torrde/miniforge/envs/moose/include
CURL_CA_BUNDLE=/Users/Shared/CAINLROOT_B64.crt
CXX=mpicxx
CXXFLAGS=-march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -fmessage-length=0 -isystem /Users/torrde/miniforge/envs/moose/include -std=c++17
CXX_FOR_BUILD=/Users/torrde/miniforge/envs/moose/bin/x86_64-apple-darwin13.4.0-clang++
F77=mpif77
F90=mpif90
F95=/Users/torrde/miniforge/envs/moose/bin/x86_64-apple-darwin13.4.0-gfortran
FC=mpif90
FC_FOR_BUILD=/Users/torrde/miniforge/envs/moose/bin/x86_64-apple-darwin13.4.0-gfortran
FFLAGS=-march=core2 -mtune=haswell -ftree-vectorize -fPIC -fstack-protector -O2 -pipe -isystem /Users/torrde/miniforge/envs/moose/include
HDF5_DIR=/Users/torrde/miniforge/envs/moose
LD=x86_64-apple-darwin13.4.0-ld
LDFLAGS=-Wl,-pie -Wl,-headerpad_max_install_names -Wl,-dead_strip_dylibs -Wl,-rpath,/Users/torrde/miniforge/envs/moose/lib -L/Users/torrde/miniforge/envs/moose/lib
LDFLAGS_LD=-pie -headerpad_max_install_names -dead_strip_dylibs -rpath /Users/torrde/miniforge/envs/moose/lib -L/Users/torrde/miniforge/envs/moose/lib
LIBMESH_DIR=/Users/torrde/miniforge/envs/moose/libmesh
MOOSE_JOBS=6
MOOSE_NO_CODESIGN=true
PATH=/Users/torrde/miniforge/envs/moose/bin:/Users/torrde/miniforge/condabin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/Users/torrde/miniforge/envs/moose/wasp/bin
PETSC_DIR=/Users/torrde/miniforge/envs/moose
REQUESTS_CA_BUNDLE=/Users/Shared/CAINLROOT_B64.crt
SSL_CERT_FILE=/Users/Shared/CAINLROOT_B64.crt
WASP_DIR=/Users/torrde/miniforge/envs/moose/wasp

##################################################################################################
Compiler(s) (CC CXX FC F77 F90):

CC=/Users/torrde/miniforge/envs/moose/bin/mpicc
CC -show:
x86_64-apple-darwin13.4.0-clang -I/Users/torrde/miniforge/envs/moose/include -I/Users/torrde/miniforge/envs/moose/include -L/Users/torrde/miniforge/envs/moose/lib -Wl,-rpath,/Users/torrde/miniforge/envs/moose/lib -Wl,-commons,use_dylibs -I/Users/torrde/miniforge/envs/moose/include -L/Users/torrde/miniforge/envs/moose/lib -lmpi -lpmpi
CC version:     clang version 12.0.1

CXX=/Users/torrde/miniforge/envs/moose/bin/mpicxx
CXX -show:
x86_64-apple-darwin13.4.0-clang++ -I/Users/torrde/miniforge/envs/moose/include -I/Users/torrde/miniforge/envs/moose/include -L/Users/torrde/miniforge/envs/moose/lib -Wl,-rpath,/Users/torrde/miniforge/envs/moose/lib -Wl,-commons,use_dylibs -I/Users/torrde/miniforge/envs/moose/include -L/Users/torrde/miniforge/envs/moose/lib -lmpicxx -lmpi -lpmpi
CXX version:    clang version 12.0.1

FC=/Users/torrde/miniforge/envs/moose/bin/mpif90
FC -show:
x86_64-apple-darwin13.4.0-gfortran -I/Users/torrde/miniforge/envs/moose/include -L/Users/torrde/miniforge/envs/moose/lib -Wl,-rpath,/Users/torrde/miniforge/envs/moose/lib -Wl,-commons,use_dylibs -I/Users/torrde/miniforge/envs/moose/include -I/Users/torrde/miniforge/envs/moose/include -L/Users/torrde/miniforge/envs/moose/lib -lmpifort -lmpi -lpmpi
FC version:     GNU Fortran (GCC) 9.5.0

F77=/Users/torrde/miniforge/envs/moose/bin/mpif77
F77 -show:
x86_64-apple-darwin13.4.0-gfortran -I/Users/torrde/miniforge/envs/moose/include -L/Users/torrde/miniforge/envs/moose/lib -Wl,-rpath,/Users/torrde/miniforge/envs/moose/lib -Wl,-commons,use_dylibs -I/Users/torrde/miniforge/envs/moose/include -I/Users/torrde/miniforge/envs/moose/include -L/Users/torrde/miniforge/envs/moose/lib -lmpifort -lmpi -lpmpi
F77 version:    GNU Fortran (GCC) 9.5.0

F90=/Users/torrde/miniforge/envs/moose/bin/mpif90
F90 -show:
x86_64-apple-darwin13.4.0-gfortran -I/Users/torrde/miniforge/envs/moose/include -L/Users/torrde/miniforge/envs/moose/lib -Wl,-rpath,/Users/torrde/miniforge/envs/moose/lib -Wl,-commons,use_dylibs -I/Users/torrde/miniforge/envs/moose/include -I/Users/torrde/miniforge/envs/moose/include -L/Users/torrde/miniforge/envs/moose/lib -lmpifort -lmpi -lpmpi
F90 version:    GNU Fortran (GCC) 9.5.0

OK

##################################################################################################
Python Sanity Checks

Verify `/usr/bin/env python3 --version` (reporting as: Python 3.10.13),
matches versions for: `which python3 && which python`

OK

##################################################################################################
Python Modules (TestHarness, run-ability)

OK

##################################################################################################
CONDA MOOSE Packages


moose-dev                 2024.05.13              build_0    https://conda.software.inl.gov/public
moose-libmesh             2024.05.05              build_0    https://conda.software.inl.gov/public
moose-libmesh-vtk         9.2.6                   build_9    https://conda.software.inl.gov/public
moose-mpich               4.0.2                  build_16    https://conda.software.inl.gov/public
moose-peacock             2023.04.11           hd610fce_0    https://conda.software.inl.gov/public
moose-petsc               3.20.3                  build_1    https://conda.software.inl.gov/public
moose-tools               2024.05.02           hf5fa97a_0    https://conda.software.inl.gov/public
moose-wasp                2024.05.08              build_0    https://conda.software.inl.gov/public

OK Conda packages and MOOSE repo/submodule versions match
```

### PR DESCRIPTION
**Pull request by TheGreatCid**
Refs #27883 

`u_old_old_old` was not being properly defined in `ExplicitDynamicsContactConstraint.h` causing certain cases of contact to fail. Past regression tests did not catch this issue as explained in #27883.

- Changed `u_old_old_old` to be initialized by `ExplicitDynamicsContactContraint`.

### CODE REVIEW

**Comment by lindsayad** [comment]
Can we change to refs #27883 instead of closes? I would really like to avoid use of this thrice older state if possible, e.g. I liked your proposed changes in https://github.com/idaholab/moose/issues/27883#issuecomment-2163769702

File: modules/contact/src/constraints/ExplicitDynamicsContactConstraint.C [hunk 1]
```diff
@@ -306,7 +307,7 @@ ExplicitDynamicsContactConstraint::solveImpactEquations(const Node & node,
 
   auto & u_dot = *_sys.solutionUDot();
   auto & u_old = _sys.solutionOld();
-  auto & u_old_old_old = _sys.solutionState(3);
+  //  auto & u_old_old_old = _sys.solutionState(3);
```
Comment (lindsayad) [inline modules/contact/src/constraints/ExplicitDynamicsContactConstraint.C:310]:
```suggestion
```

**Review by lindsayad — CHANGES_REQUESTED** [review]
It would be good to add a regression test to make sure that regressions introduced by changes like those I introduced in https://github.com/idaholab/moose/pull/27231 don't go unnoticed again in the future

**Comment by lindsayad** [comment]
> It would be good to add a regression test to make sure that regressions introduced by changes like those I introduced in #27231 don't go unnoticed again in the future

Although then again there *were* diffs on my PR in #27231 🤔 Consequently I expect this PR will also introduce diffs as well

**Comment by TheGreatCid** [comment]
A regression test would be good. I was only able to get the simulation to fail when the impact speed was sufficiently large, which I achieved by dropping the block from higher up. This takes around 2000 time steps, which is not great for a regression test.... I'm not sure how to go around this as to my knowledge there is no way to initialize velocity with explicit dynamics.

File: modules/contact/src/constraints/ExplicitDynamicsContactConstraint.C (hunk 1 above)
Comment (TheGreatCid) [reply modules/contact/src/constraints/ExplicitDynamicsContactConstraint.C]: Fixed

**Comment by lindsayad** [comment]
Indeed that goes to the meeting we had with @ddm42. Could you push on the block with a force larger than gravity to accelerate it quicker? (I'm assuming that gravity is the relevant force here since you said "dropping the block from higher")

**Comment by TheGreatCid** [comment]
> Indeed that goes to the meeting we had with @ddm42. Could you push on the block with a force larger than gravity to accelerate it quicker? (I'm assuming that gravity is the relevant force here since you said "dropping the block from higher")

That's a good point. I added a regression test using that logic. However, as expected the other explicit contact regression tests fail due to the changes in this PR. 

The question is, should I regold these tests or should I look into incorporating the changes suggested in #27883 first?

**Comment by lindsayad** [comment]
I leave that decision to you. I am not a subject matter expert in explicit dynamics. I believe you know better what a gold file should look like for this work

**Comment by TheGreatCid** [comment]
@lindsayad 
I went ahead and implemented the first order calculation of the contact displacements I suggested in #27883. It is a bit more reliable than the previous implementation. However, this is a stop gap solution as the full implementation of explicit contact still has many issues.

**Comment by lindsayad** [comment]
sounds good

**Comment by lindsayad** [comment]
@TheGreatCid can you re-push? I think the hook didn't get delivered to civet

**Comment by TheGreatCid** [comment]
@lindsayad Done.

**Comment by TheGreatCid** [comment]
It appears that the "Conda build (linux)" is failing. I'm unsure how to correct that.

**Comment by lindsayad** [comment]
@TheGreatCid looks like just the heavy test to fix

**Comment by TheGreatCid** [comment]
Should be fixed now

**Review by lindsayad — APPROVED** [review]

**Comment by lindsayad** [comment]
Thanks @TheGreatCid !

**Pull request merged**